docs(collector): update internal telemetry RPC metrics (#9554)
open-telemetry/opentelemetry.io: content/en/docs/collector/internal-telemetry.md

Reviewer: Since the RPC section was getting a little long, I suggest we eliminate the callout and create a new section about metric ownership. Very open to feedback on this idea.

```suggestion
#### Ownership of emitted metrics

Some metrics are not owned by the Collector SIG and some are limited to certain components.

**`http*`and `rpc` metrics**

These metrics are not under the Collector SIG's control, and as such, are not covered by the maturity levels below.

**`rpc` metrics**

The Collector's internal RPC metrics come from the upstream
[`otelgrpc`](https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/instrumentation/google.golang.org/grpc/otelgrpc)
instrumentation, which tracks the [OpenTelemetry RPC semantic conventions](/docs/specs/semconv/rpc/rpc-metrics/). The set of RPC
metrics emitted by the Collector has changed across releases:

| Collector version | Emitted RPC metrics                                                                                                                                                                              |
| ----------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| v0.146.x and earlier | `rpc.client.duration`, `rpc.server.duration`, `rpc.*.request.size`, `rpc.*.response.size`, `rpc.*.requests_per_rpc`, `rpc.*.responses_per_rpc`                                              |
| v0.147.0          | `rpc.client.call.duration`, `rpc.server.call.duration`, `rpc.*.request.size`, `rpc.*.response.size` (the `*_per_rpc` metrics are deprecated and no longer emitted)                              |
| v0.148.0 and later | `rpc.client.call.duration`, `rpc.server.call.duration` only                                                                                                                                    |
RPC size metrics are not emitted by
Collector v0.148.0 or later. The [RPC semantic conventions v1.40.0](https://github.com/open-telemetry/semantic-conventions/releases/tag/v1.40.0)
deprecated them due to ambiguous definitions and inconsistent implementation.

**`otelcol_processor_batch_*` metrics**

These metrics are unique to the `batchprocessor`.

**`helper` package metrics**

The `otelcol_receiver_`, `otelcol_scraper_`, `otelcol_processor_`, and `otelcol_exporter_` metrics come from their respective `helper` packages. As such, some components not using those packages might not emit them.
```

names used dots (`.`). Prior to Collector v0.120.0, internal metrics exposed
with Prometheus changed dots (`.`) to underscores (`_`) to match Prometheus
naming conventions, resulting in metric names that looked like

> `rpc_server_call_duration`.

Reviewer 2: Nitpick: This paragraph relates to behavior prior to v0.120.0, which would not follow the new semantic convention.

Versions 0.120.0 and later of the Collector use Prometheus 3.0 scrapers, so the
original `http*` and `rpc*` metric names with dots are preserved. The
[internal metrics](#lists-of-internal-metrics) on this page are listed in their
original form, such as `rpc.server.call.duration`. Because these metrics come
from instrumentation libraries, their names can change as the underlying

> semantic conventions evolve. For more information, see the

Reviewer 2: Nitpick: You already mention that metric names can change in the note below, and it seems unrelated to the subject of this paragraph, which is about dots vs. underscores.

Versions 0.120.0 and later of the Collector use Prometheus 3.0 scrapers, so the
original `http*` and `rpc*` metric names with dots are preserved. The
[internal metrics](#lists-of-internal-metrics) on this page are listed in their

> original form, such as `rpc.server.call.duration`. For more information, see the
> [Collector v0.120.0 release notes](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CHANGELOG.md#v01200).
> 

Reviewer: ```suggestion
original form, such as `rpc.server.call.duration`. For more information, see the
[Collector v0.120.0 release notes](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CHANGELOG.md#v01200).

#### RPC metrics

RPC metric names are version dependent. For instance, Collector releases
prior to v0.147.0 exposed `rpc.client.duration` and `rpc.server.duration`
instead of `rpc.client.call.duration` and `rpc.server.call.duration`.
```

Reviewer: Resolving this suggestion since RPC metrics need greater changes.

| `http.server.request.duration`                        | Measures the duration of HTTP server requests.                  | Histogram |
| `http.server.response.body.size`                      | Measures the size of HTTP server response bodies.               | Counter   |
| `otelcol_processor_batch_batch_`<br>`send_size_bytes` | Number of bytes in the batch that was sent.                     | Histogram |
| `rpc.client.call.duration`                            | Measures the duration of outbound remote procedure calls (RPC). | Histogram |
| `rpc.client.request.size`                             | Measures the size of RPC request messages (uncompressed).       | Histogram |
| `rpc.client.response.size`                            | Measures the size of RPC response messages (uncompressed).      | Histogram |
| `rpc.server.call.duration`                            | Measures the duration of inbound remote procedure calls (RPC).  | Histogram |
| `rpc.server.request.size`                             | Measures the size of RPC request messages (uncompressed).       | Histogram |

> | `rpc.server.response.size`                            | Measures the size of RPC response messages (uncompressed).      | Histogram |

Reviewer 3: AFAIK, only the *.call.* metrics are available in the latest version of collector and semconv https://opentelemetry.io/docs/specs/semconv/rpc/rpc-metrics/
```suggestion
| `rpc.client.call.duration`                            | Measures the duration of outbound remote procedure calls (RPC). | Histogram |
| `rpc.server.call.duration`                            | Measures the duration of inbound remote procedure calls (RPC).  | Histogram |
```

Reviewer 2: Indeed, I've just confirmed experimentally that:
- in 0.147.0, the RPC metrics changed to the ones currently in the PR;
- in 0.148.0, the size metrics were disabled.

Not sure why that happened in two steps, or whether there is any replacement for the size metrics?

Reviewer: I dug into this with AI's help, and the short answer is no, there is no replacement. The Semantic Conventions SIG [deprecated](https://github.com/open-telemetry/semantic-conventions/releases/tag/v1.40.0) the rpc size metrics because they had "ambiguous definitions and inconsistent implementation".

So we'll need to update this page with a little more guidance.

Thanks for pointing this out, @odubajDT!

> [!NOTE]
>
> The `http*` and `rpc*` metrics are not covered by the maturity levels below
> since they are not under the Collector SIG control.
>
> RPC metric names are version-dependent. For instance, Collector releases
> prior to 0.147.0 exposed `rpc.client.duration` and `rpc.server.duration`
> instead of `rpc.client.call.duration` and `rpc.server.call.duration`.

> >

Reviewer: I think this comment fits better in the Metric names section. See my suggestion above. As for this note, I _do_ think we should mention that the per_rpc metrics have been deprecated:

```suggestion
> All `*_per_rpc` metrics have been deprecated and are no longer exposed as of Collector v0.147.0.
>
```

Reviewer: Given all that we've learned, I'm resolving this suggestion and revising our approach.

| `http.server.response.body.size`                      | Measures the size of HTTP server response bodies.               | Counter   |
| `otelcol_processor_batch_batch_`<br>`send_size_bytes` | Number of bytes in the batch that was sent.                     | Histogram |
| `rpc.client.call.duration`                            | Measures the duration of outbound remote procedure calls (RPC). | Histogram |
| `rpc.server.call.duration`                            | Measures the duration of inbound remote procedure calls (RPC).  | Histogram |

> [!NOTE]
>
> The `http*` and `rpc*` metrics are not covered by the maturity levels below
> since they are not under the Collector SIG control.
>
> RPC metric names are version-dependent. For instance, Collector releases prior
> to 0.147.0 exposed `rpc.client.duration` and `rpc.server.duration` instead of
> `rpc.client.call.duration` and `rpc.server.call.duration`.
>
> The `otelcol_processor_batch_` metrics are unique to the `batchprocessor`.
>
> The `otelcol_receiver_`, `otelcol_scraper_`, `otelcol_processor_`, and
> `otelcol_exporter_` metrics come from their respective `helper` packages. As

> > such, some components not using those packages might not emit them.

Reviewer: ```suggestion
```

Reviewer: The new section we added below replaces this note.
